pipeline/mutate/mutator_hydrator.go

```diff
@@ -102,73 +104,86 @@ func NewMutatorHydrator(c configuration.Provider, d mutatorHydratorDependencies)
 	cache, _ := ristretto.NewCache(&ristretto.Config{
 		// This will hold about 1000 unique mutation responses.
 		NumCounters: 10000,
-		// Allocate a max of 32MB
+		// Allocate a max of 32 MB
 		MaxCost: 1 << 25,
 		// This is a best-practice value.
 		BufferItems: 64,
 	})
-	return &MutatorHydrator{c: c, d: d, client: httpx.NewResilientClientLatencyToleranceSmall(nil), hydrateCache: cache}
+	return &MutatorHydrator{c: c, d: d,
+		p: sync.Pool{
+			New: func() interface{} {
+				b := &bytes.Buffer{}
+				b.Grow(poolBufCapacityBytes)
+				return b
+			},
+		},
+		client: httpx.NewResilientClientLatencyToleranceSmall(nil), hydrateCache: cache}
 }
 
 func (a *MutatorHydrator) GetID() string {
 	return "hydrator"
 }
 
-func (a *MutatorHydrator) cacheKey(config *MutatorHydratorConfig, session string) string {
-	return fmt.Sprintf("%s|%x", config.Api.URL, md5.Sum([]byte(session)))
+// cacheKey creates a (composite) cache key.
+func (a *MutatorHydrator) cacheKey(keys ...string) string {
+	return fmt.Sprintf("%x", md5.Sum([]byte(strings.Join(keys, ""))))
 }
 
-func (a *MutatorHydrator) hydrateFromCache(config *MutatorHydratorConfig, session string) (*authn.AuthenticationSession, bool) {
-	if !config.Cache.Enabled {
-		return nil, false
-	}
-
-	item, found := a.hydrateCache.Get(a.cacheKey(config, session))
+func (a *MutatorHydrator) hydrateFromCache(key string) (*authn.AuthenticationSession, bool) {
+	item, found := a.hydrateCache.Get(key)
 	if !found {
 		return nil, false
 	}
-
 	return item.(*authn.AuthenticationSession).Copy(), true
 }
 
 func (a *MutatorHydrator) hydrateToCache(config *MutatorHydratorConfig, key string, session *authn.AuthenticationSession) {
-	if !config.Cache.Enabled {
-		return
-	}
-
-	if a.hydrateCache.SetWithTTL(a.cacheKey(config, key), session.Copy(), 0, config.Cache.ttl) {
+	if a.hydrateCache.SetWithTTL(key, session.Copy(), 0, config.Cache.ttl) {
 		a.d.Logger().Debug("Cache reject item")
 	}
 }
 
-func (a *MutatorHydrator) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, _ pipeline.Rule) error {
+func (a *MutatorHydrator) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, p pipeline.Rule) error {
 	cfg, err := a.Config(config)
 	if err != nil {
 		return err
 	}
 
-	var b bytes.Buffer
-	if err := json.NewEncoder(&b).Encode(session); err != nil {
-		return errors.WithStack(err)
-	}
+	s := a.p.Get().(*bytes.Buffer)
+	defer func() {
+		s.Reset()
+		a.p.Put(s)
+	}()
```

Comment:
I would propose to keep this PR focused on making caching possible for mutators. Any other optimizations should be done in a separate PR :)

Reply:
OK, sure, I will remove this. Even though we already know the variable will be allocated on the heap due to passing of a pointer to json decode.